mlflow/models/signature.py

```diff
@@ -130,3 +133,82 @@ def infer_signature(
     inputs = _infer_schema(model_input)
     outputs = _infer_schema(model_output) if model_output is not None else None
     return ModelSignature(inputs, outputs)
+
+
+# `t\w*\.` matches the `typing` module or its alias
+_LIST_OF_STRINGS_PATTERN = re.compile(r"^(t\w*\.)?list\[str\]$", re.IGNORECASE)
+
+
+def _is_list_str(hint_str):
+    return _LIST_OF_STRINGS_PATTERN.match(hint_str.replace(" ", "")) is not None
+
+
+_LIST_OF_STR_DICT_PATTERN = re.compile(
+    r"^(t\w*\.)?list\[(t\w*\.)?dict\[str,str\]\]$", re.IGNORECASE
+)
+
+
+def _is_list_of_string_dict(hint_str):
+    return _LIST_OF_STR_DICT_PATTERN.match(hint_str.replace(" ", "")) is not None
+
+
+def _infer_hint_from_str(hint_str):
+    if _is_list_str(hint_str):
+        return List[str]
+    elif _is_list_of_string_dict(hint_str):
+        return List[Dict[str, str]]
+    else:
+        return None
+
+
+def _get_arg_names(f):
+    return list(inspect.signature(f).parameters.keys())
+
+
+class _TypeHints:
+    def __init__(self, input_=None, output=None):
+        self.input = input_
+        self.output = output
+
+
+def _extract_type_hints(f):
+    if not hasattr(f, "__annotations__") and hasattr(f, "__call__"):
+        return _extract_type_hints(f.__call__)
+
+    if f.__annotations__ == {}:
```

Comment:
Do we need to add type hints to `mlflow.models.Model` to get this to work?

Reply:
Specifically, prior to py3.10, don't we have to access annotations on a class by doing something like this:
```
if isinstance(f, type):
  annotations = f.__dict__.get("__annotations__", None)
else:
  annotations = getattr(f, "__annotations__", None)
  ```
  
  There's some logic on this in this module note: https://docs.python.org/3/howto/annotations.html

Reply:
Do we extract type hints from `mlflow.models.Model`?